### DISCUSSION TITLE
Using send and returning context from within a promise

### DISCUSSION THREAD

**Discussion by jeffsee55** [statelyai/xstate#1698] [Q&A] [answered]
I'm building out some tooling for a GraphiQL playground and I'm using xstate/react to manage things. GraphiQL accepts a "fetcher" prop, which is a promise that lets you run a network request and return the value. In my fetcher, I want to `send({ type: "FETCH", query: args.query });` the request and await the **new context** when the machine finishes the event. Is there a way to await this? I understand that sending events is fire-and-forget, but wondering if I might use the `service` object somehow. This is where I'm at right now, but the new context values aren't populated:

```tsx
[current, send] = useMachine(MyMachine)

<GraphiQL
  fetcher={async (args: {
    query: string;
    variables: object;
    operationName: string;
  }) => {
    // this runs immediately
    send({ type: "FETCH", query: args.query });
    
    // so this isn't updated in-time
    return current.context.result
  }}
  schema={current.context.schema}
>
  ...
</GraphiQL>

```

The fetcher prop accepts a Promise:
```ts
export type Fetcher = (
  graphQLParams: FetcherParams,
  opts?: FetcherOpts,
) => Promise<FetcherResult> | Observable<FetcherResult>;
```

**Comment by jeffsee55** [comment, marked as answer]
Hmm, this works - 

```ts
fetcher={async (args: {
  query: string;
  variables: object;
  operationName: string;
}) => {
  send({ type: "FETCH", query: args.query });
  return new Promise((resolve, reject) => {
    service.onChange((context) => {
      resolve(context.result);
    });
  });
}}
```

I'll leave this here in case anyone has a better recommendation

**Reply by davidkpiano** [reply]
Looks like the right idea! Better to use `service.onTransition(({ context }) => {/* ... */})` or `service.subscribe(...)` since `.onChange` will be deprecated in v5 (redundant).
